rest/replicator_test.go

```diff
@@ -3810,3 +3810,236 @@ func TestActiveReplicatorPullConflictReadWriteIntlProps(t *testing.T) {
+	if base.GTestBucketPool.NumUsableBuckets() < 2 {
+		t.Skipf("test requires at least 2 usable test buckets")
+	}
+	if !base.TestUseXattrs() {
+		t.Skip("This test only works with XATTRS enabled")
+	}
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+
+	// startReplication starts the replication tasks and asserts that the start up
+	// operation is successful. It returns a teardown function to stop the replication
+	// that can be called in defer.
+	startReplication := func(ar *db.ActiveReplicator) func() {
+		require.NoError(t, ar.Start(), "Error starting replication")
+		return func() { require.NoError(t, ar.Stop(), "Error stopping replication") }
+	}
+
+	// requireBodyEmpty asserts that the specified document body is empty.
+	requireBodyEmpty := func(rt *RestTester, docID string) {
+		var body []byte
+		_, err := rt.Bucket().Get(docID, &body)
+		require.EqualError(t, err, fmt.Sprintf("Error during Get %s: key not found", docID))
+	}
+
+	// waitForRevision waits until the specified tombstone revision is available
+	// in the bucket backed by the specified RestTester instance.
+	waitForTombstone := func(rt *RestTester, docID, revID string) {
+		require.NoError(t, rt.WaitForCondition(func() bool {
+			doc, _ := rt.GetDatabase().GetDocument(docID, db.DocUnmarshalAll)
+			return doc.IsDeleted()
+		}))
+		doc, err := rt.GetDatabase().GetDocument(docID, db.DocUnmarshalAll)
+		require.NoError(t, err, "Error reading document from bucket")
+		require.Equal(t, revID, doc.SyncData.CurrentRev)
+		require.Empty(t, doc.Body(), "Document body should be empty")
+	}
+
+	// requireBodyEmpty asserts that the specified document body is not empty.
+	requireBodyNotEmpty := func(rt *RestTester, docID, revID, bodyValue string) {
+		doc, err := rt.GetDatabase().GetDocument(docID, db.DocUnmarshalAll)
+		require.NoError(t, err, "Error reading document from bucket")
+		require.Equal(t, revID, doc.SyncData.CurrentRev)
+		require.NotEmpty(t, doc.Body(), "Document body shouldn't be empty")
+		require.Equal(t, bodyValue, doc.GetDeepMutableBody()["key"])
+	}
+
+	// createOrUpdateDoc creates a new document or update an existing document with the
+	// specified document id, revision id and body value in a channel named "alice".
+	createOrUpdateDoc := func(rt *RestTester, docID, revID, bodyValue string) string {
+		body := fmt.Sprintf(`{"key":%q,"channels":["alice"]}`, bodyValue)
+		dbURL := "/db/" + docID
+		if docID != "" {
+			dbURL = "/db/" + docID + "?rev=" + revID
+		}
+		resp := rt.SendAdminRequest(http.MethodPut, dbURL, body)
+		assertStatus(t, resp, http.StatusCreated)
+		require.NoError(t, rt.WaitForPendingChanges())
+		return respRevID(t, resp)
+	}
+
+	defaultConflictResolverWithTombstoneTests := []struct {
+		// A unique name to identify the unit test.
+		name string
+
+		// bodyValue is used to control the digest value for revision generation.
+		bodyValue string
+
+		// revGenHigh is used to make a revision tie by updating the remote doc.
```

Comment:
revGenHigh and revGenTie could be simplified as 'remoteRevGen', which I think would also make it clearer in the test definition what the scenario is, instead of having to dig into the test code to understand what the updated generation turns out to be on the remote

Reply:
When looking at this further, it makes more sense to pass the body values as test input for each document update mutation on remote and by that way we can easily control the higher revision generation, revision tie, lower/higher revision digest etc. on the the remote (rt2). Refactored the code in latest revision.